classes/local/store/s3/client.php

```diff
@@ -89,29 +93,56 @@ public function __wakeup() {
      * @return bool
      */
     protected function is_functional() {
-        return isset($this->client);
+        return !empty($this->client);
     }
 
     /**
-     * Check if the client configured properly.
-     *
-     * @param \stdClass $config Client config.
-     * @return bool
+     * Tests that the configuration is ok.
+     * @return result
      */
-    protected function is_configured($config) {
-        if (empty($config->s3_bucket)) {
-            return false;
-        }
+    public function test_configuration(): result {
+        $configcheck = $this->check_configuration($this->config);
+        $details = '';
+
+        $lookup = [
+            true => get_string('settings:config:exists', 'tool_objectfs'),
+            false => get_string('settings:config:missing', 'tool_objectfs'),
+            null => get_string('settings:config:na', 'tool_objectfs'),
+        ];
 
-        if (empty($config->s3_region)) {
-            return false;
+        foreach ($configcheck as $check => $result) {
+            $details .= $check . ":" . $lookup[$result] . "\n";
         }
 
-        if (empty($config->s3_usesdkcreds) && (empty($config->s3_key) || empty($config->s3_secret))) {
-            return false;
-        }
+        $valid = $this->is_configuration_valid($this->config);
+        $status = $valid ? result::OK : result::ERROR;
+        $summary = $valid ? get_string('check:configuration:ok', 'tool_objectfs')
+            : get_string('check:configuration:error', 'tool_objectfs');
+        $details = nl2br($details);
 
-        return true;
+        return new result($status, $summary, $details);
+    }
+
+    /**
+     * Checks the configuration is valid.
+     * @return array key => value where the value is true if ok, false if bad, or null if n/a
+     */
+    protected function check_configuration($config) {
+        return [
```

Comment:
Should this check more than just these - i.e. maybe the credentials check in here as well ?